src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library.cljs

```diff
@@ -0,0 +1,155 @@
+           optional (clojure.set/difference (-> props :library-schema :propertyOrder set) required)]
+       {:required-props required
+        :optional-props optional}))
+   :render
+   (fn [{:keys [this]}]
+     [modal/OKCancelForm
+      {:header "Define Dataset Attributes"
+       :content (react/create-element (react/call :form this))
+       :ok-button #(react/call :on-ok this)}])
+   :form
+   (fn [{:keys [props state]}]
+     (let [library-schema (:library-schema props)]
+       [:div {:style {:width "50vw"}}
+        [:div {:style {:maxHeight 400 :overflowY "auto"
+                       :padding "0.5em" :border style/standard-line
+                       :marginBottom "1em"}}
+         (map (fn [property-key]
+                (let [{:keys [hidden title inputHint enum]} (get-in library-schema [:properties property-key])
+                      required? (contains? (:required-props @state) property-key)
+                      pk-str (name property-key)]
+                  (when-not hidden
+                    [:div {}
+                     (style/create-form-label (str title (when required? " (required)")))
+                     (cond enum (style/create-identity-select {:ref pk-str} enum)
+                           required? [input/TextField {:ref pk-str
+                                                       :style {:width "100%"}
+                                                       :placeholder inputHint
+                                                       :predicates [(when required? (input/nonempty pk-str))]}]
+                           :else (style/create-text-field {:ref (name property-key)
+                                                           :style {:width "100%"}
+                                                           :placeholder inputHint}))])))
+              (calculate-display-properties library-schema))]
+        (style/create-validation-error-message (:validation-error @state))]))
+   :on-ok
+   (fn [{:keys [props state refs]}]
+     (let [library-schema (:library-schema props)
+           field-data (doall
+                        (map (fn [property-key]
+                               (let [{:keys [hidden enum]} (get-in library-schema [:properties property-key])
+                                     pk-str (name property-key)
+                                     required? (contains? (:required-props @state) property-key)]
+                                 [property-key
+                                  {:required? required?
+                                   :value (not-empty
+                                            (cond hidden (resolve-hidden property-key (:workspace props))
+                                                  enum (common/get-text refs pk-str)
+                                                  required? (do (input/validate refs pk-str)
+                                                              (input/get-text refs pk-str))
```

Comment:
in other words - could common/get-text and input/get-text ever return different values, when passed the same refs/id? If they *could* return different values, then if a PO changes any given field from required to optional, this code would calculate a different value.

common/get-text and input/get-text have duplicated code internally - but not complete copies of each other. Could they be more centralized?